wgengine/magicsock/magicsock.go

```diff
@@ -3291,6 +3313,46 @@ func (c *Conn) UpdateStatus(sb *ipnstate.StatusBuilder) {
 	})
 }
 
+// updateStats updates the statistics counters with the src, dst, and cnts.
+// It is a noop if statistics gathering is not enabled.
+func (c *Conn) updateStats(src netip.Addr, dst netip.AddrPort, cnts tunstats.Counts) {
+	// This method is written to be inlineable.
+	if c.stats.enabled.Load() {
+		c.updateStatsEnabled(src, dst, cnts)
+	}
+}
+
+func (c *Conn) updateStatsEnabled(src netip.Addr, dst netip.AddrPort, cnts tunstats.Counts) {
+	conn := flowtrack.Tuple{Src: netip.AddrPortFrom(src, 0), Dst: dst}
+	c.stats.mu.Lock()
+	defer c.stats.mu.Unlock()
+	if c.stats.m == nil {
+		c.stats.m = make(map[flowtrack.Tuple]tunstats.Counts)
+	}
+	c.stats.m[conn] = c.stats.m[conn].Add(cnts)
+}
+
+// SetStatisticsEnabled enables per-connection packet counters.
+// Disabling statistics gathering does not reset the counters.
+// ExtractStatistics must be called to reset the counters and
+// be periodically called while enabled to avoid unbounded memory use.
+func (c *Conn) SetStatisticsEnabled(enable bool) {
+	c.stats.enabled.Store(enable)
+}
+
+// ExtractStatistics extracts and resets the counters for all active connections.
+// It must be called periodically otherwise the memory used is unbounded.
+//
+// The source is always a peer's tailscale IP address, while the destination
+// is the peer's phyiscal IP address and port.
```

Comment:
The same peer. I added another sentence to explicitly call out that the current node's address is not relevant here.

Reply:
So is this kinda an abuse of the flowtrack.Tuple type? It might have the right shape, but the wrong names/docs?

Reply:
In a sense, yes.

My original prototype had this return `map[key.NodePublic]tunstats.Counts` and the caller would do the work to resolve the `key.NodePublic` back into a `netip.Addr`.

Having this return a `flowtrack.Tuple` makes it such that `tstun.Wrapper.ExtractStatistics` and `magicsock.Conn.ExtractStatistics` have the same signature, so that the caller (in `wgengine/netlog`) can operate on them in a similar way.